ssl/quic/quic_impl.c

```diff
@@ -9,179 +9,2104 @@
+ * ===============================================
+ *
+ * Here we dispatch calls to the steady-state front-end I/O API functions; that
+ * is, the functions used during the established phase of a QUIC connection
+ * (e.g. SSL_read, SSL_write).
+ *
+ * Each function must handle both blocking and non-blocking modes. As discussed
+ * above, all QUIC I/O is implemented using non-blocking mode internally.
+ *
+ *   (BIO/)SSL_read                 => ossl_quic_read
+ *   (BIO/)SSL_write                => ossl_quic_write
+ *
+ */
+
+static int blocking_mode(QUIC_CONNECTION *qc)
 {
-    return -1;
+    return 1; // TODO
+}
+
+/*
+ * SSL_get_error {{{2
+ * -------------
+ */
+int ossl_quic_get_error(const QUIC_CONNECTION *qc, int i)
+{
+    return SSL_ERROR_NONE; // TODO
+}
+
+/*
+ * SSL_write {{{2
+ * ---------
+ */
+struct quic_write_again_args {
+    QUIC_CONNECTION *qc;
+    const void      *buf;
+    size_t          len;
+    size_t          *written;
+};
+
+static int quic_write_again(void *arg)
+{
+    struct quic_write_again_args *args = arg;
+
+    // TODO what if connection is torn down due to error while blocking (e.g. rx
+    // bad frame)
+
+    if (!ossl_quic_sstream_append(args->qc->stream0->sstream, args->buf,
+                                   args->len, args->written))
+        return -2;
+
+    if (*args->written > 0)
+        /* appended at least one byte, the SSL_write op can finish now */
+        return 1;
+
+    return 0; /* did not write anything, keep trying */
 }
 
 int ossl_quic_write(SSL *s, const void *buf, size_t len, size_t *written)
 {
-    BIO *wbio = SSL_get_wbio(s);
-    int ret;
-    SSL_CONNECTION *sc = SSL_CONNECTION_FROM_QUIC_SSL(s);
+    QUIC_CONNECTION *qc = QUIC_CONNECTION_FROM_SSL(s);
+    struct quic_write_again_args args;
+    int res;
 
-    if (sc == NULL || wbio == NULL)
+    *written = 0;
+
+    if (!EXPECT_QUIC_CONN(qc))
         return 0;
 
-    sc->rwstate = SSL_WRITING;
-    ret = BIO_write_ex(wbio, buf, len, written);
-    if (ret > 0 || !BIO_should_retry(wbio))
-        sc->rwstate = SSL_NOTHING;
-    return ret;
+    // TODO if closed
+    // TODO stop_sending, reset_stream?
+
+    if (qc->stream0 == NULL || qc->stream0->sstream == NULL)
+        return 0;
+
+    if (!ossl_quic_sstream_append(qc->stream0->sstream, buf,
+                                  len, written)) {
+        /* Stream already finished or allocation error. */
+        *written = 0;
+        return 0;
+    }
+
+    if (*written > 0) {
+        /*
+         * We have appended at least one byte to the stream.
+         * Potentially mark stream as active, depending on FC.
+         */
+        ossl_quic_stream_map_update_state(&qc->qsm, qc->stream0);
+
+        /*
+         * Try and send.
+         *
+         * TODO(QUIC): It is probably inefficient to try and do this
+         * immediately, plus we should eventually consider Nagle's algorithm
+         * etc.
+         */
+        reactor_tick(&qc->rtor);
```

Comment:
The logic here looks unexpected to me. If we're in blocking mode we shouldn't return until everything is written. But here we don't even check if we're in blocking mode if we've written at least one byte, and then we just return. If we're not in blocking mode I expect SSL_write to fail if we have failed to write *all* the data (and SSL_get_error() should return SSL_ERROR_WANT_WRITE) - unless SSL_MODE_ENABLE_PARTIAL_WRITE has been set.

In the case of non-blocking mode and SSL_MODE_ENABLE_PARTIAL_WRITE has not been set, then we if we *did* do a partial write then we will need to remember how much we wrote, and continue from where we left off next time we are called.

Reply:
This behaviour of SSL_write is highly surprising.

As an aside, are our our own API docs accurate? SSL_write(3) talks about this, but BIO_write(3) says nothing about partial writes not being allowed. But BIO_write(3) for BIO_f_ssl is a simple wrapper around SSL_write(3). BIO_write(3) doesn't say anything about having to pass the same buffer again. So the BIO docs seem to be quite misleading here.

Reply:
I don't think BIO_write() generally has this restriction but SSL_write does. BIO_write() for BIO_f_ssl() will just never do a partial write I think - which is still within the API contract for BIO_write().